tools/clang/unittests/HLSLExec/ExecutionTest.cpp

```diff
@@ -13236,6 +13205,173 @@ void ExecutionTest::WaveSizeTest() {
   }
 }
 
+bool TestShaderRangeAgainstRequirements(UINT shaderminws, UINT shadermaxws,
+                                        UINT minws, UINT maxws) {
+  if (shaderminws > maxws) {
+    return false;
+  }
+  if (shadermaxws < minws) {
+    return false;
+  }
+  return true;
+}
+
+void RunWaveSizeRangeTest(UINT minWaveSize, UINT maxWaveSize,
+                          std::shared_ptr<st::ShaderOpSet> ShaderOpSet,
+                          CComPtr<ID3D12Device> pDevice,
+                          dxc::DxcDllSupport &m_support) {
+  // format shader source
+  const char waveSizeTestShader[] =
+      "struct TestData { \r\n"
+      "  uint count; \r\n"
+      "}; \r\n"
+      "RWStructuredBuffer<TestData> data : register(u0); \r\n"
+      "\r\n"
+      "// Note: The 3 WAVESIZE variables below will be defined via compiler "
+      "option -D\r\n"
+      "[wavesize(MINWAVESIZE, MAXWAVESIZE, PREFWAVESIZE)]\r\n"
+      "[numthreads(" stringify(MAX_WAVESIZE) "*2,1,1)]\r\n"
+                                             "void main(uint3 tid : "
+                                             "SV_DispatchThreadID ) { \r\n"
+                                             "  data[tid.x].count = "
+                                             "WaveGetLaneCount(); \r\n"
```

Comment:
What's the value of writing the WaveGetLaneCount() value from every thread?  It should be a constant value from the device, and could be written to one location by one thread ` if (WaveIsFirstLane())`.  Same for the single-value test.